File: ObserverPattern/ObserverPattern.cpp
#include <iostream>
#include "string.h"
using namespace std;
#define MAX_SUBSCRIBERS 100
//IObserver and IObservable are the crux of the program. Rest are customisable.
class IObserver
{
  public:
  virtual void update(const string articleName) = 0;
};
class IObservable
{
  public:
  virtual void subscribe(IObserver* o) = 0;
  virtual void unsubscribe(IObserver* o) = 0;
  virtual void notifySubsribers() = 0;
};

class MobileUser : public IObserver
{
  private:
  string userName="DEFAULT";
  
  public:
  MobileUser(const string userName)
  {
    this->userName = userName;
  }
  void update(const string articleName)
  {
    cout<<"Trending now - \""<<articleName<<"\". Viewed by "<<userName<<" on Mobile.\n";
  }
};

class LaptopUser : public IObserver
{
  private:
  string userName="DEFAULT";
  
  public:
  LaptopUser(const string userName)
  {
    this->userName = userName;
  }
  void update(const string articleName)
  {
    cout<<"Trending now - \""<<articleName<<"\". Viewed by "<<userName<<" on Laptop.\n";
  }
};
class NewsWebsite : public IObservable
{
  private:
  string latestTrendingArticle="DEFAULT";
  IObserver* subList[MAX_SUBSCRIBERS];
  int subCount = 0;
  
  public:
  void subscribe(IObserver* o)
  {
    if(subCount<MAX_SUBSCRIBERS)
    {
      subList[subCount++] = o;
    }
  }
  void unsubscribe(IObserver* o)
  {
    //too lazy to write.
  }
  void notifySubsribers()
  {
    for(int ind=0;ind<subCount;ind++)
    {
      subList[ind]->update(latestTrendingArticle);
    }
  }
  void changeArticle(string articleName)
  {
    latestTrendingArticle = articleName;
    notifySubsribers();
  }
};

int main()
{
  NewsWebsite news;
  MobileUser mUser1("Chaitanya");
  LaptopUser lUser1("Ayush");
  
  news.subscribe(&mUser1);
  news.changeArticle("Sachin Tendulkar calls off retirement. Mixed reactions!!");
  
  news.subscribe(&lUser1);
  news.changeArticle("Karen dials 911 on a white boy selling lemonade. Cops puzzled.");
  return 0;
}

File: ObserverPattern.java
import java.util.ArrayList;

interface IObserver{
  void update(String data);
}
interface IObservable{
  void add(IObserver o);
  void remove(IObserver o);
  void sendNotify();
}

class MobileUser implements IObserver
{
  String userName;
  public MobileUser(String userName)
  {
    this.userName = userName;
  }
  public void update(String data)
  {
    System.out.println("Trending now - \""+ data+"\". Viewing on mobile by "+userName+".");
  }
}

class LaptopUser implements IObserver
{
  String userName;
  public LaptopUser(String userName)
  {
    this.userName = userName;
  }
  public void update(String data)
  {
    System.out.println("Trending now - \""+ data+"\". Viewing on laptop by "+userName+".");
  }
}

class NewsWebsite implements IObservable
{
  ArrayList<IObserver> subscriberList = new ArrayList<IObserver>();
  String latestTrendingArticle = "DEFAULT";
  public void add(IObserver o)
  {
    subscriberList.add(o);
  }
  public void remove(IObserver o)
  {
    subscriberList.remove(o);
  }
  public void sendNotify()
  {
    for(int i=0;i<subscriberList.size();i++)
    {
      subscriberList.get(i).update(latestTrendingArticle);
    }
  }
  
  public void changeTrendingArticle(String articleName)
  {
    latestTrendingArticle = articleName;
    sendNotify();
  }
}
public class ObserverPattern{
  
  public static void main(String[] args)
  {
    MobileUser mUser1 = new MobileUser("Chaitanya");
    LaptopUser lUser1 = new LaptopUser("Ayush");
    NewsWebsite instNewsWebsite = new NewsWebsite();
    
    instNewsWebsite.add(mUser1);
    instNewsWebsite.changeTrendingArticle("Valve announces Half Life 3. Fans elated!!");
    
    instNewsWebsite.add(lUser1);
    instNewsWebsite.changeTrendingArticle("Salman Khan apologises for the black buck killing.");
  }
  
}